### ISSUE TITLE
Index migration to 7.17.1 doesnt create _meta field for .watches and .triggered_watches indices

### ISSUE THREAD

**Issue by TheRiffRafi** [elastic/elasticsearch#84880]
### Elasticsearch Version

7.17.1

### Java Version

_bundled_

### OS Version

Any OS

### Problem Description

Upgrade to 7.17.1 fails to create mappings for the entire _meta object in the .watches and .triggered_watches indices.
So far we have only confirmed the behavior upgrading from version 7.13.1

### Steps to Reproduce

So far it seems reproduction is not consistent, unsure on exact cluster settings or state needed to generate the problem. 
We have confirmed from 4 users however that this happened just after upgrading to 7.17.1

### Logs (if relevant)

Missing _meta field in mapping [_doc] of index [.triggered_watches], assuming mappings update required
Missing _meta field in mapping [_doc] of index [.watches], assuming mappings update required

**Comment by TheRiffRafi** [comment]
Current work-around is to add the field and a value below the current running version.
For example if the cluster is running version 7.17.1 then set 7.17.0 (or the version before upgrade if known) for the version value:

```json
PUT .watches/_mapping
{
    "_meta": {
      "version" : "7.17.0"
  }
}

PUT .triggered_watches/_mapping
{
    "_meta": {
      "version" : "7.17.0"
  }
} 
```

**Comment by BLKingLLC** [comment]
I had the same issue with 8.1, I tried the workaround from @TheRiffRafi, the updates went through just fine but then I'm left with ```update_mapping [_doc]``` with no other context and the watcher service still not running.  Any idea what's going on?

**Comment by williamrandolph** [comment]
I looked into this and managed to replicate it with the following steps:

1. Run Elasticsearch 6.8.20 with a trial license
2. Create a watch (the one from [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/watcher-getting-started.html) is fine
3. Upgrade to 7.13.1. There will be a warning about failing to notify a cluster state updater, with a stack trace containing `java.lang.IllegalStateException: Cannot read version string in index .watches`
4. Upgrade to 7.17.1. There will be many warning messages saying `Missing _meta field in mapping [doc] of index [.watches], assuming mappings update required

Now that I've reproduced this, I don't think this actually has to do with the Feature Migration code, since I didn't have to run a feature migration to reproduce. I believe the bug has to do with the `SystemIndexManager`, which is supposed to handle mapping updates behind the scenes for internal system indices.

**Comment by williamrandolph** [comment]
It looks like we added some handling for a missing `_meta` field in #83896, which is why we see different messages in 7.17.1 for the same issue.

**Comment by williamrandolph** [comment]
After digging into this code a bit, I found a bug in the system index definitions for Watcher.

First, the thing that has been confusing me. The code in 7.17.x is supposed to quietly restore a missing _meta field in mappings, so these warnings are supposed to disappear soon after the index goes green.

However, before repairing mappings, our system index manager checks to see that the system index has the correct format. There are two ways that we could find a "wrong format":

1. The index's `index.format` setting does not match the index format number hardcoded in the system index settings.
2. The index's mapping `type` is incorrect, which, as far as I can tell, means that it is something other than `_doc`.

The bug (which I will fix) has to do with the first problem. The hard-coded index format number is incorrect in the watcher system index definition.

The way I've been reproducing the issue brings up both of these problems. The second problem occurs because the watcher mapping type in 6.8.x is `doc`, rather than `_doc`. This can only be fixed by reindexing.

If you're on 7.17, the system feature migration can clean all of this up for you.

```
$ curl -s -XPOST 'localhost:9200/_migration/system_features'
{"accepted":true,"features":[{"feature_name":"watcher"}]}

$ curl -s -XGET 'localhost:9200/_migration/system_features
# lots of output, just make sure that the overall migration status is `NO_MIGRATION_NEEDED` before proceeding

$ curl -s -XGET 'localhost:9200/.watches/_mappings' | jq '.'
{
  ".watches-reindexed-for-8": {
    "mappings": {
      "dynamic": "strict",
      "_meta": {
        "version": "7.17.2"
      },
      "properties": {
        [...]
      }
    }
  }
}
```

In short, running a system feature migration should get rid of the warning.

It might be worth putting in a counter to make sure this message only gets logged once. It looks pretty serious when it shows up so much.

**Comment by thateriksson** [comment]
Thanks @williamrandolph does this work for version 8.X? So could we do the workaround mentioned by @TheRiffRafi  and the above to clean the warnings?

**Comment by dolaru** [comment]
@thateriksson I was able to use `POST _migration/system_features` in 8.1.1 to fix the warnings.

**Comment by williamrandolph** [comment]
I need to update the advice I gave: `POST _migration/system_features` will probably create problems with the geoip ingest plugin, so I don't advise doing it. Now I would suggest manually adding the `_meta.version` fields to the mappings. See https://github.com/elastic/sdh-elasticsearch/issues/5736#issuecomment-1072532797 for details and links.

**Comment by lucabelluccini** [comment]
It seems the common factor for this problem to occur is the presence of a legacy index template of any kind.
It is enough to have a match all index template such as the following one to trigger the problem:
```
  "sometemplate" : {
    "order" : 0,
    "index_patterns" : [
      ".*"
    ],
    "settings" : {
      "index" : {
        "routing" : {
          "allocation" : {
            "require" : {
              "data" : "hot"
            }
          }
        }
      }
    },
    "mappings" : { },
    "aliases" : { }
  },
```

**Comment by ldecrescenzo** [comment]
Hi team! I have a customer with the same issue, they already tried @TheRiffRafi workaround and the problem remains with the message:
```
“[2023-06-29T13:47:44,048][WARN ][o.e.p.PersistentTasksClusterService] [jefsrv] persistent task upgrade-system-indices failed

java.lang.IllegalStateException: unable to create new index [.triggered_watches-reindexed-for-8] because it would match composable template [.triggered_watches]”
```

Is there something else we can try in this case?
